lib/spack/docs/module_file_support.rst

```diff
@@ -0,0 +1,704 @@
+     one. This allows a site to build the same libraries or applications against different
+     implementations of ``mpi`` and ``lapack``, and let LMod switch safely from one to the
+     other.
+
+.. warning::
+  Deep hierarchies and ``lmod spider``
+   For hierarchies that are deeper than three layers ``lmod spider`` may have some issues.
+   See `this discussion on the LMod project <https://github.com/TACC/Lmod/issues/114>`_.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Filter out environment modifications
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Modifications to certain environment variables in module files are generated by
+default, for instance by prefix inspections in the default configuration file.
+There are cases though where some of these modifications are unwanted.
+Suppose you need to avoid having ``CPATH`` and ``LIBRARY_PATH``
+modified by your ``dotkit`` modules:
+
+.. code-block:: yaml
+
+   modules:
+     dotkit:
+       all:
+         filter:
+           # Exclude changes to any of these variables
+           environment_blacklist: ['CPATH', 'LIBRARY_PATH']
+
+The configuration above will generate dotkit module files that will not contain
+modifications to either ``CPATH`` or ``LIBRARY_PATH`` and environment module
+files that instead will contain these modifications.
+
+^^^^^^^^^^^^^^^^^^^^^
+Autoload dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+In some cases it can be useful to have module files directly autoload their dependencies.
+This may be the case for Python extensions, if not activated using ``spack activate``:
+
+.. code-block:: yaml
+
+   modules:
+     tcl:
+       ^python:
+         autoload: 'direct'
+
+The configuration file above will produce module files that will automatically
+load their direct dependencies. The allowed values for the  ``autoload`` statement
+are either ``none``, ``direct`` or ``all``.
+
+.. note::
+  TCL prerequisites
+     In the ``tcl`` section of the configuration file it is possible to use
+     the ``prerequisites`` directive that accepts the same values as ``autoload``.
+     It will produce module files that have a ``prereq``
+     statement instead of automatically loading other modules.
+
+=======================
+Module file maintenance
+=======================
+
+Spack not only provides great flexibility in the generation of module files
+and in the customization of both their layout and content, but also ships with
+a tool to ease the burden of their maintenance in production environments.
+This tool is the ``spack module`` command:
+
+.. code-block:: console
+
+  $ spack module --help
+  usage: spack module [-h] SUBCOMMAND ...
+
+  positional arguments:
+    SUBCOMMAND
+      refresh   Regenerate module files
+      find      Find module files for packages
+      rm        Remove module files
+      loads     Prompt the list of modules associated with a constraint
+
+  optional arguments:
+    -h, --help  show this help message and exit
```

Comment:
For this help message and the ones below, you should replace the entire thing with:
```rst
.. command-output:: spack module --help
```
This will automatically print the output of the command, so it will always be correct, even if someone adds a new option.